### PR TITLE
[ALLUXIO-3045] Add external minicluster

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-3045

This PR adds the external minicluster and applies it to JournalShutdownIntegrationTest to replace a couple of our codebase's flakiest tests.

Features of the external cluster:
- Can be configured with any number of masters and workers, as well as any Alluxio properties.
- Cluster uses a temporary directory as its work directory. An `alluxio-site.properties` is written to `workdir/conf`, the underfs is `workdir/underFSStorage`, the worker ramdisks are `workdir/ramdisk${worker_id}`, and the journal folder is `workdir/journal`. Master logs are written to `logs-master${master_id}`, and worker logs are written to `logs-worker${worker_id}`
- The cluster supports a `saveWorkDir` method which will create a tarball of worker state and store it in `target/artifacts`, where Jenkins will temporarily archive it. This will be invaluable for debugging failing tests.

Efforts taken to avoid leaking external processes:
- Register a JVM shutdown hook to destroy the external processes on JVM exit
- If a test uses the external cluster as a `@Rule`, the cluster will automatically get destroyed when the test ends.
- Tests which use the external cluster directly should use `try ... finally` to ensure that they call `cluster.destroy()` when they are done.
- External processes run in a wrapper which will cause them to terminate after 20 minutes.

### CODE REVIEW

File: core/common/src/test/java/alluxio/ConfigurationTestUtils.java [hunk 1]
```diff
@@ -89,9 +89,9 @@ public static void resetConfiguration() {
     conf.put(PropertyKey.MASTER_JOURNAL_TAILER_SHUTDOWN_QUIET_WAIT_TIME_MS, "50ms");
     conf.put(PropertyKey.MASTER_JOURNAL_TAILER_SLEEP_TIME_MS, "10ms");
 
-    // If tests fail to connect they should fail early rather than using the default ridiculously
-    // high retries
-    conf.put(PropertyKey.USER_RPC_RETRY_MAX_NUM_RETRY, "3");
+    // To keep tests fast, we should do more retries with a lower max wait time.
+    conf.put(PropertyKey.USER_RPC_RETRY_MAX_NUM_RETRY, "60");
+    conf.put(PropertyKey.USER_RPC_RETRY_MAX_SLEEP_MS, "500ms");
```
Comment (calvinjia) [inline core/common/src/test/java/alluxio/ConfigurationTestUtils.java]: Unrelated to this PR, we have deprecated the use of `_MS` and other time unit suffixed properties

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+  /** Closer for closing all resources that must be closed when the cluster is destroyed. */
+  private Closer mCloser;
+  private List<Master> mMasters;
+  private List<Worker> mWorkers;
+  /** Addresses of all masters. Should have the same size as {@link #mMasters}. */
+  private List<MasterNetAddress> mMasterAddresses;
+  private State mState;
+  private TestingServer mCuratorServer;
+
+  private ExternalCluster(Map<PropertyKey, String> properties, int numMasters, int numWorkers,
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/external/ExternalCluster.java]: External doesn't convey to me what this cluster is doing, maybe `MultiProcess` or `Standalone`?

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+      String clusterName) {
+    mProperties = properties;
+    mNumMasters = numMasters;
+    mNumWorkers = numWorkers;
+    // Add a unique number so that different runs of the same test use different cluster names.
+    mClusterName = clusterName + ThreadLocalRandom.current().nextLong();
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/external/ExternalCluster.java]: What's the significance of `clusterName`, could it be purely randomly generated?

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    if (zkEnabled()) {
+      mCuratorServer = mCloser
+          .register(new TestingServer(-1, AlluxioTestDirectory.createTemporaryDirectory("zk")));
+      mProperties.put(PropertyKey.ZOOKEEPER_ADDRESS, mCuratorServer.getConnectString());
+    } else {
+      mProperties.put(PropertyKey.MASTER_HOSTNAME, mMasterAddresses.get(0).getHostname());
+      mProperties.put(PropertyKey.MASTER_RPC_PORT,
+          Integer.toString(mMasterAddresses.get(0).getRpcPort()));
+      mProperties.put(PropertyKey.MASTER_WEB_PORT,
+          Integer.toString(mMasterAddresses.get(0).getWebPort()));
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/external/ExternalCluster.java]: make `mMasterAddress.get(0)` a named variable?

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            mMasterAddresses));
+  }
+
+  /**
+   * @return a client for interacting with the cluster
+   */
+  public synchronized FileSystem getFileSystemClient() {
+    Preconditions.checkState(mState == State.STARTED,
+        "must be in the started state to get an fs client, but state was %s", mState);
+    MasterInquireClient inquireClient = getMasterInquireClient();
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/external/ExternalCluster.java]: We cannot use the factory method, will we have problems if we `reset` later?

File: minicluster/src/main/java/alluxio/external/Master.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+   * @param mWorkDir the work directory to use for the master process
+   * @param address the address information for the master
+   * @param masterId an ID for this master, used to distinguish it from other masters in the same
+   *        cluster
+   */
+  public Master(File mWorkDir, int masterId, MasterNetAddress address) throws IOException {
+    mLogsDir = new File(mWorkDir, "logs-master" + masterId);
+    mConfDir = new File(mWorkDir, "conf");
+    mOutFile = new File(mLogsDir, "master.out");
+    mAddress = address;
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/external/Master.java]: Explicitly pass in these values instead of generating them?

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java [hunk 7]
```diff
@@ -90,28 +93,63 @@ public final void before() throws Exception {
...
+        .setClusterName("singleMasterJournalStopIntegration")
+        .setNumWorkers(0)
+        .setNumMasters(1)
+        .build();
+    try {
+      cluster.start();
+      FileSystem fs = cluster.getFileSystemClient();
+      runCreateFileThread(fs);
+      cluster.killPrimaryMaster();
+      cluster.startMaster(0);
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: Should we have a different API for just bringing up the cluster afterward?

File: minicluster/src/main/java/alluxio/external/Master.java (hunk 6 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/external/Master.java]: done

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java (hunk 7 above)
Comment (aaudiber) [reply tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: To keep things simple, the cluster doesn't keep track of which processes are running or destroyed - this management is left to individual tests. We could have add a "startAll" method which loops starts all masters and workers, but I'd prefer to be explicit in tests about what is being restarted.

File: core/common/src/test/java/alluxio/ConfigurationTestUtils.java (hunk 1 above)
Comment (aaudiber) [reply core/common/src/test/java/alluxio/ConfigurationTestUtils.java]: 👍

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java (hunk 2 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/external/ExternalCluster.java]: changed to `MultiProcess`

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java (hunk 3 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/external/ExternalCluster.java]: it's useful for identifying which test the cluster is for - in particular when we collect artifacts for failed clusters, this name is used in the tarball name.

File: minicluster/src/main/java/alluxio/external/ExternalCluster.java (hunk 4 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/external/ExternalCluster.java]: done

**Comment by aaudiber** [comment]
Jenkins, test this please

**Comment by aaudiber** [comment]
Jenkins, test this please

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java (hunk 7 above)
Comment (calvinjia) [reply tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: We provide the API `killPrimaryMaster`, and the caller has no idea which master is killed, should that return the master index?

**Comment by calvinjia** [comment]
@aaudiber Thanks for the update, had a question left about the api.

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java (hunk 7 above)
Comment (aaudiber) [reply tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: Good idea, added

**Comment by aaudiber** [comment]
jenkins, test this please

**Comment by aaudiber** [comment]
@apc999 All tests should be consistently passing now

File: minicluster/pom.xml [hunk 8]
```diff
@@ -46,6 +46,11 @@
       <artifactId>curator-test</artifactId>
       <scope>compile</scope>
     </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>compile</scope>
```
Comment (apc999) [inline minicluster/pom.xml]: why compile scope rather than test scope?

Comment (aaudiber) [reply minicluster/pom.xml]: This way, ~MultiProcessCluster` (In `src/`, similar to `LocalAlluxioCluter`) can extend `TestRule`. I think this makes it more convenient to use

File: minicluster/src/main/java/alluxio/multi/process/ExternalProcess.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    args.add(mClazz.getCanonicalName());
+    ProcessBuilder pb = new ProcessBuilder(args);
+    pb.redirectError(mOutFile);
+    pb.redirectOutput(mOutFile);
+    mProcess = pb.start();
+  }
+
+  @Override
+  public synchronized void close() {
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/ExternalProcess.java]: shall this method be `stop` rather than `close`?

File: minicluster/src/main/java/alluxio/multi/process/Worker.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+  /**
+   * @return the worker's web port
+   */
+  public int getWebPort() {
+    return mWebPort;
+  }
+
+  @Override
+  public synchronized void close() {
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/Worker.java]: shall this be `stop`? do we ever use this worker as a closable to take advantage from `close()`?

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java [hunk 11]
```diff
@@ -90,28 +93,65 @@ public final void before() throws Exception {
...
   @Test
   public void singleMasterJournalStopIntegration() throws Exception {
-    LocalAlluxioCluster cluster = setupSingleMasterCluster();
-    runCreateFileThread(cluster.getClient());
-    System.out.println(Configuration.get(PropertyKey.MASTER_JOURNAL_FOLDER));
-    // Shutdown the cluster
-    cluster.stopFS();
-    CommonUtils.sleepMs(TEST_TIME_MS);
-    awaitClientTermination();
-    reproduceAndCheckState(mCreateFileThread.getSuccessNum());
```
Comment (apc999) [inline tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: don't you need to remove this method, since it is no more used?

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            CreateFileOptions.defaults().setBlockSizeBytes(100));
+        break;
+      } catch (Exception e) {
+        // This can indicate that the worker hasn't connected yet, so we must delete and retry.
+        if (System.currentTimeMillis() < timeoutMs) {
+          if (fs.exists(new AlluxioURI(testFile))) {
+            fs.delete(new AlluxioURI(testFile));
+          }
+        } else {
+          Assert.fail(String.format("Timed out trying to create a file. Latest exception: %s",
```
Comment (apc999) [inline minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: nit: `fail(...)`

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+public final class MultiProcessClusterTest {
+  private boolean mSuccess;
+  private MultiProcessCluster mCluster;
+
+  @Rule
+  public Timeout mTimeout = Timeout.millis(Constants.MAX_TEST_DURATION_MS);
+
+  @Before
+  public void before() {
+    mSuccess = false;
```
Comment (apc999) [inline minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: nit: why not `private boolean mSuccess = false` to save `before()`?

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+  /**
+   * Destroys the test cluster, saving its work directory in case of failure.
+   *
+   * This cannot be done with @After because @After methods are not necessarily called when a test
+   * times out due to a Timeout rule.
+   */
+  private void cleanup() throws Exception {
+    if (mCluster != null) {
+      if (!mSuccess) {
+        mCluster.saveWorkdir();
```
Comment (apc999) [inline minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: shall we log anything here? how are we expected to know it is not successful?

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+  @Test
+  public void zookeeper() throws Exception {
+    mCluster = MultiProcessCluster.newBuilder()
+        .setClusterName("zookeeper")
+        .addProperty(PropertyKey.ZOOKEEPER_ENABLED, "true")
+        .setNumMasters(3)
+        .setNumWorkers(2)
+        .build();
+    boolean success = false;
```
Comment (apc999) [inline minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: is this var ever used any where?

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 16]
```diff
@@ -0,0 +1,426 @@
...
+import java.util.Map.Entry;
+import java.util.concurrent.ThreadLocalRandom;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Class for starting, stopping, and interacting with an Alluxio cluster where each master and
+ * worker runs in its own process.
+ *
+ * Compared to {@link LocalAlluxioCluster}, {@link MultiProcessCluster} is
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: shall we actually rename `LocalAlluxioCluster` to `SingleProcessCluster` then? leave this up to you (and can be in different PRs)

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 17]
```diff
@@ -0,0 +1,426 @@
...
+  private final Map<PropertyKey, String> mProperties;
+  private final int mNumMasters;
+  private final int mNumWorkers;
+  private final String mClusterName;
+
+  /** Base directory for storing configuration and logs. */
+  private File mWorkDir;
+  /** Closer for closing all resources that must be closed when the cluster is destroyed. */
+  private Closer mCloser;
+  private List<Master> mMasters;
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: final

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+  private List<Worker> mWorkers;
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: final

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+
+  private final Map<PropertyKey, String> mProperties;
+  private final int mNumMasters;
+  private final int mNumWorkers;
+  private final String mClusterName;
+
+  /** Base directory for storing configuration and logs. */
+  private File mWorkDir;
+  /** Closer for closing all resources that must be closed when the cluster is destroyed. */
+  private Closer mCloser;
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: final

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 20]
```diff
@@ -0,0 +1,426 @@
...
+
+  /**
+   * Kills the primary master.
+   *
+   * If no master is currently primary, this method blocks until a primary has been elected, then
+   * kills it.
+   *
+   * @return the ID of the killed master
+   */
+  public synchronized int killPrimaryMaster() {
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: nit: `killPrimaryMasterOrWait`?

File: minicluster/src/main/java/alluxio/multi/process/ExternalProcess.java (hunk 9 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/ExternalProcess.java]: done

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 16 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: This touches a lot of code, so I'd prefer to do it in a separate PR after merging `external-minicluster` to master

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 19 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: done

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 17 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: done

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 18 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: done

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 20 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: changed name to `waitForAndKillPrimaryMater`

File: minicluster/src/main/java/alluxio/multi/process/Worker.java (hunk 10 above)
Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/Worker.java]: We register all masters and workers with `mCloser` in `MultiProcessCluster` so that they can easily be destroyed at the end. If we change this to `stop`, we would need an extra adaptor class to use with the `Closer`, so I prefer calling implementing `close` for simplicity.

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java (hunk 13 above)
Comment (aaudiber) [reply minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: donen

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java (hunk 15 above)
Comment (aaudiber) [reply minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: removed, thanks - it was left over from a previous refactor

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java (hunk 14 above)
Comment (aaudiber) [reply minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: If it wasn't successful, it must have thrown an exception which will be captured and displayed by junit

File: minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java (hunk 12 above)
Comment (aaudiber) [reply minicluster/src/test/java/alluxio/multi/process/MultiProcessClusterTest.java]: done

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java (hunk 11 above)
Comment (aaudiber) [reply tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: good catch, done

**Comment by aaudiber** [comment]
@apc999 Thanks for the careful review, I've updated the PR to reflect your comments

File: minicluster/pom.xml (hunk 8 above)
Comment (apc999) [reply minicluster/pom.xml]: ack

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java (hunk 16 above)
Comment (apc999) [reply minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: sg

File: minicluster/src/main/java/alluxio/multi/process/Worker.java (hunk 10 above)
Comment (apc999) [reply minicluster/src/main/java/alluxio/multi/process/Worker.java]: ack. I think then it makes more sense to have our "closer" implementation that can also takes `Server<T>` in addition to `Closeable` as argument. What do you think --- we can handle this in later PRs

Comment (aaudiber) [reply minicluster/src/main/java/alluxio/multi/process/Worker.java]: I don't have a strong preference, but I'd still lean towards making things that should be cleaned up `Closeable` so that they can be used with `try-with-resources`, `Closer`, and resource leak detection tools
